packages/jupyter-ai/src/components/chat-input.tsx

```diff
@@ -26,8 +39,122 @@ type ChatInputProps = {
   sx?: SxProps<Theme>;
 };
 
+type SlashCommandOption = {
+  id: string;
+  label: string;
+  description: string;
+};
+
+/**
+ * List of icons per slash command, shown in the autocomplete popup.
+ *
+ * This list of icons should eventually be made configurable. However, it is
+ * unclear whether custom icons should be defined within a Lumino plugin (in the
+ * frontend) or served from a static server route (in the backend).
+ */
+const DEFAULT_SLASH_COMMAND_ICONS: Record<string, JSX.Element> = {
+  ask: <FindInPage />,
+  clear: <HideSource />,
+  export: <Download />,
+  generate: <MenuBook />,
+  help: <Help />,
+  learn: <School />,
+  unknown: <MoreHoriz />
+};
+
+/**
+ * Renders an option shown in the slash command autocomplete.
+ */
+function renderSlashCommandOption(
+  optionProps: React.HTMLAttributes<HTMLLIElement>,
+  option: SlashCommandOption
+): JSX.Element {
+  const icon =
+    option.id in DEFAULT_SLASH_COMMAND_ICONS
+      ? DEFAULT_SLASH_COMMAND_ICONS[option.id]
+      : DEFAULT_SLASH_COMMAND_ICONS.unknown;
+
+  return (
+    <li {...optionProps}>
+      <Box sx={{ lineHeight: 0, marginRight: 4, opacity: 0.618 }}>{icon}</Box>
+      <Box sx={{ flexGrow: 1 }}>
+        <Typography
+          component="span"
+          sx={{
+            fontSize: 'var(--jp-ui-font-size1)'
+          }}
+        >
+          {option.label}
+        </Typography>
+        <Typography
+          component="span"
+          sx={{ opacity: 0.618, fontSize: 'var(--jp-ui-font-size0)' }}
+        >
+          {' — ' + option.description}
+        </Typography>
+      </Box>
+    </li>
+  );
+}
+
 export function ChatInput(props: ChatInputProps): JSX.Element {
-  function handleKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
+  const [slashCommandOptions, setSlashCommandOptions] = useState<
+    SlashCommandOption[]
+  >([]);
+
+  /**
+   * Effect: fetch the list of available slash commands from the backend on
+   * initial mount to populate the slash command autocomplete.
+   */
+  useEffect(() => {
+    async function getSlashCommands() {
+      const slashCommands = (await AiService.listSlashCommands())
+        .slash_commands;
+      setSlashCommandOptions(
+        slashCommands.map<SlashCommandOption>(slashCommand => ({
+          id: slashCommand.slash_id,
+          label: '/' + slashCommand.slash_id + ' ',
+          description: slashCommand.description
+        }))
+      );
+    }
+    getSlashCommands();
+  }, []);
+
+  // whether any option is highlighted in the slash command autocomplete
+  const [highlighted, setHighlighted] = useState<boolean>(false);
+
+  // controls whether the slash command autocomplete is open
+  const [open, setOpen] = useState<boolean>(false);
+
+  /**
+   * Effect: Open the autocomplete when the user types a slash into an empty
+   * chat input. Close the autocomplete and reset the last selected value when
+   * the user clears the chat input.
+   */
+  useEffect(() => {
+    if (props.value === '/') {
```

Comment:
If I tab out of the input box with a partly typed slash command, then reenter it, I don't see the auto-completer popup. Replacing this with a regex, like testing if `props.value` matches `/\/\w*/` (slash plus 0 or more word characters), might fix this.